The joy of slow programming (#1)
adamniedzielski/blog: _posts/2021-09-17-the-joy-of-slow-programming.md

---
layout: post

> title: "The joy of slow programming"

Reviewer: Referencing https://en.wikipedia.org/wiki/Zen_and_the_Art_of_Motorcycle_Maintenance you could call it "Zen and the Art of Side Project Maintenance" :)
Might get you some unexpected clicks :)
(But I totally get it if you want "slow programming" in the title, it's a phrase worth coining!)

Author: I read this book 😄 (or at the very least started reading), because I remember the title.

Yes, I'd like to establish the phrase "slow programming", similarly to "slow food" 😉

Reviewer 2: I think "Slow programming" and "boring code" are really good examples of [concept handles](https://notes.andymatuschak.org/z5vA4vw86DKNq22xt6pRWhumeRmSzwV6hxRHE):

> Successful concept handles can really amplify a vague idea which many people sort of understand but can’t point to and talk about. If you give that vague notion a crisp, catchy name, you can unlock a lot of conversation and reflection

Reviewer: Just realized your title already IS a pop culture reference, whether intentional or not: to "The Joy of Painting" with Bob Ross :smile:

Author: I didn’t do that consciously, but the phrase “the joy of X” just sounded good to me 😅

Reviewer: "Concept handle" is a concept handle too :repeat: :slightly_smiling_face:

the implementation, added tests soon, but not immediately, and configured CI
after I had a running version in production.

I was working at a slow, steady pace, trying to enjoy the every moment of it.
As I removed the pressure of time from myself, I could actually enjoy the
little details like a [new Rubocop rule that I didn't know before](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ConstantDefinitionInBlock)
or read up on [libraries that I was considering to use](https://github.com/bblimke/webmock).

After a few days of working I had the first version and could actually start
sending notifications to my friend. This meant that I was receiving feedback
early and could adjust the application based on it, without making too many
false assumptions.

Was the fast "hacking evening" with Python a wasted time? No, not really.
Thanks to this session I learned that Telegram is a messaging app where it's
extremely easy to send messages programmatically and how to create my own
Telegram bot. From the time perspective I think about it as prototyping session,
although during the actual session I didn't have this prototyping mindset.

> ### Next boring steps

Reviewer: Amazing headline (in this world of click-bait and attention-grabbing everything, I absolutely mean it) :sparkles: :)

Author: Haha, boring can be good, like boring Ruby code 😉

I always started by copy pasting the previous scraper and the tests for it.
The code was repetitive but I avoided reaching for any early abstractions.
The classes were simple and could fit in a single screen. The actual business
logic - getting information from the HTML using Nokogiri and regular expressions
- was unique to each company website; with very few common elements.

This approach paid off as soon as I found the first website loading data
dynamically from an unsecured JSON API :) Because of the lack of any
authentication it was just easier to access the JSON API directly. My
abstraction was lightweight so I didn't have to modify it. I just wrote this
one particular scraper differently.

At some point I introduced the Apartment model to the application. At this
stage I wasn't yet sure what data I can actually get in a consistent way. The
only thing that I knew for sure was that I need to store a unique identifier for
each apartment to avoid sending duplicated notifications after each subsequent
check. That's why I added a mandatory `external_id` field. Everything else ended
up in a JSONB `properties` field. This was simple and allowed me to move fast

> with delivering business value while moving slow with technical assumptions.

Reviewer: ```suggestion
with delivering value while moving slow with technical assumptions.
```
I know what you mean by "business value", just wouldn't use it here since it's nothing business-related, and too many things in this world are business-related :)

### Starting to use the app myself

A month or two have passed and guess what? I found myself in a need for a new
apartment. Yes, I had to move out from my flat share and find a new apartment
for myself. The application that I originally developed for my friend became my

> own hope on the cruel Berlin flat market.

Reviewer: Personally I'd make the wording a bit less dramatic, bc the actual problem is way more horrible than it sounds here, but it's just some info floating around without really being adequately contextualized by the article, since this is not what the article even tries to do, if you know what I mean

Author: You're right, the actual problem is hard to describe to anybody who doesn't have this experience, so the dramatic wording here is out of context.

However, there were about 30-40 housing associations that I was interested in.

I started with a long list of links and every day was going through all of them,
mostly to find out that no new offers appeared. This was a tedious and
frustrating process. So I began migrating the sources to be automatically handled
by my app, one by one. From the technical point of view they were no different
than the previous sources, so I didn't have to change anything in my assumptions.
Again, I was working step by step, no more than one hour per day. Each migrated
source meant one link less to click every morning, the list was shrinking, and
the progress was visible to me. This kept me motivated.

### Happy news

One day I received a notification from a housing cooperative that was
particularly appealing to me. The apartment was in a neighbourhood close to my
old apartment and a really good deal for the money. I immediately applied for
it and from that moment on things moved fast. Next day I had a viewing and
within a week I signed the contract.

The application that I developed as a side project scored me a new apartment!

> That's my most successful side project so far.

Reviewer: ```suggestion
That means it's my most successful side project so far.
```
(Oral vs. written EN)

I attribute this success to the philosophy of "slow programming" that I followed:
1. put a hard limit on the time per day that you invest into the side project
2. move in small steps, don't rush, and appreciate the little learnings that
come every day

> 3. stay motivated by making a tiny but visible improvement with each step

Reviewer: I feel it would be nice to add a final thing like "Happy (slow-)coding!" or sth. like that, just to round it off :)

Author: Great idea!

somehow know which apartment offers are new and which are not, in order to avoid
sending the same notifications over and over again. I tried to get the code
running on the provided infrastructure, with each passing minute getting more
frustrated that it doesn't work yet, and more disconnected from the reality.

While I was getting deeper and deeper into the programming zone, my friend was
sitting next to me, getting more and more bored, because at some point I
stopped explaining what I'm currently doing. Not exactly the "hacking evening"
anyone would wish for. After my frustration reached a certain level and my energy
dropped to zero I did `rm -rf the-project` and told my friend: "Well, I guess
that I need to think about it more in the next days".

### New approach - take it slow

The result of this thinking were a few guiding ideas that I decided on:
1. use ["boring" technologies](/blog/2018/11/22/boring-ruby-code/) that I already
know - Ruby, Nokogiri, Rails, and Heroku.
2. no more than one hour of work on the project per day
3. focus on delivering the smallest possible thing that actually provides some
value to my friend

> 4. strive for simplicity from the technical perspective and not perfection

Reviewer 2: It's interesting that you're contrasting "technical simplicity" against "perfection" as if they're opposed: for me technical simplicity only comes at the end of a refactoring/thinking process, and I mentally associate that process with my own perfectionism 😁

Author: Hmm... interesting point! 😉 I guess that I meant "first simple thing that works". That doesn't always yield overall architectural simplicity. I'll see if I can re-write the point to make that clearer and still keep it short and crisp 😄

I thought about them as defining my new mindset of "slow programming". And this
is how [Neue Wohnung](https://github.com/adamniedzielski/neue_wohnung) was born.
If you look at the commits in the reverse order you can see that I started with
the implementation, added tests soon, but not immediately, and configured CI
after I had a running version in production.

> I was working at a slow, steady pace, trying to enjoy the every moment of it.

Reviewer 2: "trying to enjoy the every moment of it" --> "trying to enjoy every moment"

Reviewer 2: (I really like this idea btw - been trying to do this myself)

As I removed the pressure of time from myself, I could actually enjoy the
little details like a [new Rubocop rule that I didn't know before](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ConstantDefinitionInBlock)

> or read up on [libraries that I was considering to use](https://github.com/bblimke/webmock).

Reviewer 2: considering to use ---> considering using.

After a few days of working I had the first version and could actually start
sending notifications to my friend. This meant that I was receiving feedback
early and could adjust the application based on it, without making too many
false assumptions.

Was the fast "hacking evening" with Python a wasted time? No, not really.

> Thanks to this session I learned that Telegram is a messaging app where it's

Reviewer 3: ```suggestion
Thanks to this session I learned that Telegram is a messaging app for which it's
```

extremely easy to send messages programmatically and how to create my own

> Telegram bot. From the time perspective I think about it as prototyping session,

Reviewer 3: ```suggestion
Telegram bot. From the time perspective I think about it as a prototyping session,
```

although during the actual session I didn't have this prototyping mindset.

### Next boring steps

The scraper was working in production (I quickly set up Heroku) and my friend was
receiving new apartment offers every day. The next step was very boring - add
support for more companies. It was a tedious and repetitive activity to write
this code, but I was never doing it for more than an hour per day and at the
same time immediately getting the business value out of it - more and more
apartment offers were reaching my friend - and this kept me motivated.

I always started by copy pasting the previous scraper and the tests for it.

> The code was repetitive but I avoided reaching for any early abstractions.

Reviewer 3: ```suggestion
The code was repetitive but I avoided reaching for any early [abstractions](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction).
```

The classes were simple and could fit in a single screen. The actual business
logic - getting information from the HTML using Nokogiri and regular expressions
- was unique to each company website; with very few common elements.

> This approach paid off as soon as I found the first website loading data

Reviewer 2: I don't understand how the logical connection you're trying to make between this paragraph and the previous one - in what way does "discovering that data was loaded dynamically" mean that your approach was good?

Author: The connection is following: there are no early abstractions so it's easier to adapt the code to a new situation where we parse JSON and not HTML. I'll see how I can make that clearer.

dynamically from an unsecured JSON API :) Because of the lack of any
authentication it was just easier to access the JSON API directly. My
abstraction was lightweight so I didn't have to modify it. I just wrote this
one particular scraper differently.

At some point I introduced the Apartment model to the application. At this

> stage I wasn't yet sure what data I can actually get in a consistent way. The

Reviewer 2: can -> could

> only thing that I knew for sure was that I need to store a unique identifier for

Reviewer 2: need -> needed

each apartment to avoid sending duplicated notifications after each subsequent
check. That's why I added a mandatory `external_id` field. Everything else ended
up in a JSONB `properties` field. This was simple and allowed me to move fast
with delivering value while moving slow with technical assumptions.

### Starting to use the app myself

> A month or two have passed and guess what? I found myself in a need for a new

Reviewer 3: ```suggestion
A month or two had passed and guess what? I found myself in a need for a new
```
